src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD201CancelAfterSwitchToMainThreadAnalyzer.cs

```diff
@@ -0,0 +1,112 @@
+                    var cancellationTokenNoneSymbol = cancellationTokenTypeSymbol.GetMembers(nameof(CancellationToken.None)).Single();
+                    startCtxt.RegisterOperationAction(Utils.DebuggableWrapper(c => this.AnalyzeInvocation(c, switchMethods, cancellationTokenTypeSymbol, cancellationTokenNoneSymbol)), OperationKind.Invocation);
+                }
+            });
+        }
+
+        private void AnalyzeInvocation(OperationAnalysisContext context, ImmutableArray<ISymbol> switchMethods, INamedTypeSymbol cancellationTokenTypeSymbol, ISymbol cancellationTokenNoneSymbol)
+        {
+            var invocationOperation = (IInvocationOperation)context.Operation;
+            if (invocationOperation.TargetMethod is object && switchMethods.Contains(invocationOperation.TargetMethod))
+            {
+                IArgumentOperation? tokenArgument = invocationOperation.Arguments.FirstOrDefault(arg => arg.Value?.Type?.Equals(cancellationTokenTypeSymbol) ?? false);
+
+                // Filter out missing arguments, `default` and `default(CancellationToken)`
+                if (tokenArgument is null || tokenArgument.IsImplicit || tokenArgument.Value.IsImplicit || tokenArgument.Value is IDefaultValueOperation)
+                {
+                    return;
+                }
+
+                // Did CancellationToken.None get passed in?
+                if (tokenArgument.Value is IPropertyReferenceOperation p && p.Member.Equals(cancellationTokenNoneSymbol))
+                {
+                    return;
+                }
+
+                // Did the invocation get followed by a check on that CancellationToken?
+                var statement = Utils.FindAncestor<IExpressionStatementOperation>(invocationOperation);
+                if (statement?.Parent is IBlockOperation containingBlock)
+                {
+                    int currentStatement = containingBlock.Operations.IndexOf(statement);
+                    IOperation? nextOperation = containingBlock.Operations.Length > currentStatement + 1 ? containingBlock.Operations[currentStatement + 1] : null;
+                    if (!IsTokenCheck(nextOperation, tokenArgument.Value))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, invocationOperation.Syntax.GetLocation()));
+                    }
+                }
+                else
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, invocationOperation.Syntax.GetLocation()));
+                }
+
+                bool IsTokenCheck(IOperation? consideredStatement, IOperation token)
+                {
+                    // if (token.IsCancellationRequested)
+                    if (consideredStatement is IConditionalOperation ifStatement &&
+                        ifStatement.Condition is IMemberReferenceOperation memberAccess &&
+                        Utils.IsSameSymbol(memberAccess.Instance, token) &&
```

Comment:
@mavasani do we have a good way to check "is reference to same object" which accounts for nested property accesses (e.g. both are references to X.Y.Z)?

Reply:
If we have one, I can fix and stop skipping the `CancellationToken_ThrowWithCTFromWrongObject_ProducesDiagnostic` test.